backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/query/impl/LuceneSearchScrollImpl.java

```diff
@@ -85,7 +85,7 @@ public void close() {
 		}
 
 		// no more results check
-		if ( scrollIndex >= search.totalHitCount() ) {
+		if ( search.totalHitCount() == null || scrollIndex >= search.totalHitCount() ) {
```

Comment:
Yes, in case of scroll the count is always required to be exact (no disbled - no threshold).
Even Though in case of a limit fetching timeout the hit count can be a lower bound.
That is the case tested by `SearchQueryScrollIT#all_truncateAfter`, in which we got a lower bounded hit count.
I think that for this very case we should not rely on hit count to skip the extraction and go over instead, 
so I would change it. What do you think?

Reply:
Ok. So you want to take the total hit count threshold into account for scrolls. Not sure it's absolutely necessary, but let's try.

Yes, if you want to allow setting a threshold, we should not rely on the total hit count here.

However, we can't just skip this condition: that's basically the terminal condition of user loops when scrolling, which will switch `SearchScrollResult#hasHits()` to `false`.

I think you should mirror what you did for Elasticsearch: don't rely on the total hit count to determine whether it's the last scroll result, but rather on the number of extracted results (before loading), i.e. `hasHits() == !LuceneLoadableSearchResult.extractedData.isEmpty()`. Then you can get rid of the use of totalHitCount here entirely.

By the way, maybe we should always set the totalHitCountThreshold to `0` for scrolls, since we won't use the total hit count at all?
And maybe we should do the same for Elasticsearch, set `track_total_hits` to `false` for scrolls?

Reply:
I think that currently even the Elasticsearch backend uses total hit count to skip the hit extraction if, see `Elasticsearch7SearchResultExtractor#extract` row 65:
``` java
final List<Object> extractedHits = hitCount > 0 ?
  extractHits( extractContext ) : Collections.emptyList();
```
So probably, even if indirectly, `ElasticsearchLoadableSearchResult#hasHits` depends on total hit count too.

Anyway, I agree with you that there is no reason to use hit count to skip, in case it is exactly zero, the extraction. It is better to use the current next/page hitSize. E.g.:

``` java
public boolean hasHits() {
  TopDocs topDocs = luceneCollectors.getTopDocs();
  return ( topDocs == null ) ? false : topDocs.scoreDocs.length > 0;
}
```

Reply:
> I think that currently even the Elasticsearch backend uses total hit count to skip the hit extraction if this is 0, see Elasticsearch7SearchResultExtractor#extract row 65:
> 
> ```
> final List<Object> extractedHits = hitCount > 0 ?
>   extractHits( extractContext ) : Collections.emptyList();
> ```

Not really, this is just an optimization. If the hit count is 0, `extractHits()` would have returned an empty list anyway, so we skip it.